ciao-controller/internal/datastore/datastore.go

```diff
@@ -1406,6 +1406,29 @@ func (ds *Datastore) AddBlockDevice(device types.BlockData) error {
 	return nil
 }
 
+func (ds *Datastore) DeleteBlockDevice(ID string) error {
+	// lock both tenants and devices maps
+	var err error
+
+	ds.bdLock.Lock()
+	ds.tenantsLock.Lock()
+
```

Comment:
These locks aren't commented in ciao-controller/internal/datastore/datastore.go so I'm not 100% sure the design of what they're meant to protect, but I think you have introduced a race below...